### PR TITLE
opt: add cost of buffering rows in the cost model

### PR DESCRIPTION
**Pull request by rytaft**
**opt: fix oversight in costing of hash joins**

This commit fixes an oversight in the costing of hash joins in which
equality conditions were given a higher cost in hash joins than in other
types of joins.

Release note: None

**opt: add cost of buffering rows in the cost model**

This commit adds a new function in the cost model called `rowBufferCost`,
which adds a cost for buffering rows according to a ramp function:
```
                 cost
                factor

                   |               spillRowCount
  spillCostFactor _|                  ___________ _ _ _
                   |                 /
                   |                /
                   |               /
               0  _| _ _ _________/______________________    row
                   |                                        count
                        noSpillRowCount
```
This function models the fact that operators that buffer rows become more
expensive the more rows they need to buffer, since eventually they will need
to spill to disk. The exact number of rows that cause spilling to disk varies
depending on a number of factors that we don't model here. Therefore, we use
a ramp function rather than a step function to account for the uncertainty
and avoid sudden surprising plan changes due to a small change in stats.

`rowBufferCost` is currently called in four places: when costing sorts, hash
joins, hash group bys, and hash set operations.

Release note (performance improvement): The optimizer is now less likely to
create query plans that require buffering a large number of rows in memory.
This can improve performance by reducing memory pressure and reducing the
likelihood that execution operators will need to spill to disk.

### CODE REVIEW

**Comment by rytaft** [comment]
There were two TPC-H plans that changed:
```
      time before      time after       delta
Q3     6.77s ± 0.18s    7.13s ± 0.16s   +5.4%
Q18   12.44s ± 0.40s   12.15s ± 0.06s   -2.3%
```
If we're not comfortable in the slight increase for Q3 I can tweak the model a bit more -- the `n0` and `k` parameters can be tuned.

**Review by mgartner — APPROVED** [review]
Very cool! :lgtm: 

> If we're not comfortable in the slight increase for Q3 I can tweak the model a bit more -- the n0 and k parameters can be tuned.

It looks like we're now adding a sort under the group by so that we can do a streaming group by instead of a hash group by. This seems odd to me - if the sort were to spill to disk, then wouldn't it be likely the the group by also spills to disk? In which case, adding a sort seems suboptimal compared to doing a hash group by without the sort. Do you think that using this model for hash set operations might fix this? It seems like it would only make hash set operations more expensive, and so it wouldn't fix this, but I'm not sure.

> Reviewed 12 of 12 files at r1, 14 of 14 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66559#-:-McMXCCYEAtpO1HyT33N:bya21yn)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/xform/coster.go, line 1181 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66559#-McMSSXn67VtqFUTP26u:-McMSSXn67VtqFUTP26v:b-dyaiqa) ([raw file](https://github.com/cockroachdb/cockroach/blob/7259a1f3f35582a706ccf6d0bdcd066ad0ad1247/pkg/sql/opt/xform/coster.go#L1181)):*
> ```
> https://en.wikipedia.org/wiki/Logistic_function
> ```

Was there some literature that hinted that this function would be useful here, or was this just sitting in the back of your head waiting for an application?

You can add "In computer science: database query optimization" in the "Applications" section of this article. :laughing: 
___
*[pkg/sql/opt/xform/testdata/coster/groupby, line 28 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66559#-McMWHvz22FPzNLD7lFN:-McMWHvz22FPzNLD7lFO:b-a4u2ue) ([raw file](https://github.com/cockroachdb/cockroach/blob/7259a1f3f35582a706ccf6d0bdcd066ad0ad1247/pkg/sql/opt/xform/testdata/coster/groupby#L28)):*
> ```
> 
> # As the rows increase, the hash groupby should become more expensive according
> # to a logistic function.
> ```

nit: explain what these test results are supposed to show. Maybe something like: "The group by's individual cost (the difference in the group by's cost and the group by's input cost) follows a logistic curve. The difference in the group by's individual cost between 100k and 150k rows is greater than the differences between 50k and 100k rows, and 150k and 200k rows." :shrug: 

I threw the numbers into a spreadsheet and plotted the cost to verify that they do indeed follow a logistic function curve. It would be nice to make it easier to verify these test results, but I don't see any way to do that.

**Review by RaduBerinde — APPROVED** [review]
Nice!

> It looks like we're now adding a sort under the group by ..

For Q3, note that the added sort is not directly under the group-by, it is under a lookup join and sorts a smaller row set (221623 vs 492693).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66559#-:-McORp5bAgTOou09cxvv:bjwqwr6)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rytaft)

___
*[pkg/sql/opt/xform/coster.go, line 1202 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66559#-McOTULYAxkH-2Fkdf2f:-McOTULYAxkH-2Fkdf2g:bshrs29) ([raw file](https://github.com/cockroachdb/cockroach/blob/7259a1f3f35582a706ccf6d0bdcd066ad0ad1247/pkg/sql/opt/xform/coster.go#L1202)):*
> ```Go
> func (c *coster) rowBufferCost(rowCount float64) memo.Cost {
> 	var factor memo.Cost
> 	if rowCount > 1 {
> ```

[nit] would be easier to follow if we just did `if rowCount <= 1 { return 0 }`
___
*[pkg/sql/opt/xform/coster.go, line 1205 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66559#-McOU3um-ntOh6J6Gng_:-McOU3um-ntOh6J6Gnga:b4b9r4m) ([raw file](https://github.com/cockroachdb/cockroach/blob/7259a1f3f35582a706ccf6d0bdcd066ad0ad1247/pkg/sql/opt/xform/coster.go#L1205)):*
> ```Go
> 		const (
> 			L  = seqIOCostFactor
> 			n0 = 100000
> ```

We try to have constants at the top of the file for such parameters. And we should try to give some intuition about them. For example, `n0` is a guess for a "cutoff" where disk spilling occurs. 
`k` relates to the "width" of the sigmoid but the actual value is not intuitive. We should maybe define it indirectly. A couple of ideas:
 - define two "cutoffs", one where the factor scaling is 0.01 and one where the scaling is 0.99 (so effectively where the transition begins and ends) and calculate `n0` and `k` from those
 - define the midpoint `n0` and a percentage X% that indicates that anything that's more than X% away from the midpoint is either <= 0.01 or >= 0.99.

The chosen value of `k` seems super "tight" to me.. For 95000 rows we get approximately no penalty and for 105000 we get the full penalty. I think the transition should be much wider (maybe `0.0001` or even `0.00005`

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66559)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rytaft)

___
*[pkg/sql/opt/xform/coster.go, line 1204 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66559#-McOaDEf2UCNn_dhCPMQ:-McOaDEf2UCNn_dhCPMR:bswcgle) ([raw file](https://github.com/cockroachdb/cockroach/blob/7259a1f3f35582a706ccf6d0bdcd066ad0ad1247/pkg/sql/opt/xform/coster.go#L1204)):*
> ```Go
> 	if rowCount > 1 {
> 		const (
> 			L  = seqIOCostFactor
> ```

[nit] maybe include a comment explaining how to think about seqIOCostFactor vs randIOCostFactor here. The reality is that in most cases we're not simply appending rows, we're inserting into a sorted structure. But there is inherent batching in the LSM tree, so seqIOCostFactor still seems like the better choice.

**Comment by RaduBerinde** [comment]
___
*[pkg/sql/opt/xform/coster.go, line 1205 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66559#-McOU3um-ntOh6J6Gng_:-McPg6TECPu0j5fB27KD:b-2r45ap) ([raw file](https://github.com/cockroachdb/cockroach/blob/7259a1f3f35582a706ccf6d0bdcd066ad0ad1247/pkg/sql/opt/xform/coster.go#L1205)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We try to have constants at the top of the file for such parameters. And we should try to give some intuition about them. For example, `n0` is a guess for a "cutoff" where disk spilling occurs. 
`k` relates to the "width" of the sigmoid but the actual value is not intuitive. We should maybe define it indirectly. A couple of ideas:
 - define two "cutoffs", one where the factor scaling is 0.01 and one where the scaling is 0.99 (so effectively where the transition begins and ends) and calculate `n0` and `k` from those
 - define the midpoint `n0` and a percentage X% that indicates that anything that's more than X% away from the midpoint is either <= 0.01 or >= 0.99.

The chosen value of `k` seems super "tight" to me.. For 95000 rows we get approximately no penalty and for 105000 we get the full penalty. I think the transition should be much wider (maybe `0.0001` or even `0.00005`
</blockquote></details>

It might be easier to reason about if we just made it a linear ramp between a "low" and a "high" cutoff, which would be a pretty smooth transition if there is significant difference between the cutoffs. It would be much cheaper to compute too.

**Comment by jordanlewis** [comment]
This is awesome! Do any of the changed TPCH plans have observable performance differences?

**Review by rytaft — COMMENTED** [review]
TFTRs!

I've changed to use a ramp function as @raduberinde suggested, and made it a much more gradual increase. As a result, TPC-H Q3 went back to the old plan. So the only plan that has changed is Q18, which seems to be an improvement over the previous plan. The benchmark results (copied from above) are:
```
      time before      time after       delta
Q18   12.44s ± 0.40s   12.15s ± 0.06s   -2.3%
```

I've put all the changes in a new commit, but if they look ok, I'll squash with the previous commit before merging.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66559#-:-McOw4cl5zySDUB0_mnn:by8ih7s)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner)

___
*[pkg/sql/opt/xform/coster.go, line 1181 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66559#-McMSSXn67VtqFUTP26u:-McOwpQB2ozK1947MVL2:b-op1sim) ([raw file](https://github.com/cockroachdb/cockroach/blob/7259a1f3f35582a706ccf6d0bdcd066ad0ad1247/pkg/sql/opt/xform/coster.go#L1181)):*
> Was there some literature that hinted that this function would be useful here, or was this just sitting in the back of your head waiting for an application?

I didn't look to see if other people were using it for this type of cost model, but I think it's common to use it as a sort of "smooth step function". The sigmoid function used for neural nets was what first popped into my head, but eventually I settled on the logistic function, which seemed to have the right knobs to tune.

> You can add "In computer science: database query optimization" in the "Applications" section of this article. 😆

:smile: 
___
*[pkg/sql/opt/xform/coster.go, line 1202 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66559#-McOTULYAxkH-2Fkdf2f:-McPWsby4TZFgs7gXP9u:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7259a1f3f35582a706ccf6d0bdcd066ad0ad1247/pkg/sql/opt/xform/coster.go#L1202)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] would be easier to follow if we just did `if rowCount <= 1 { return 0 }`
</blockquote></details>

Done.
___
*[pkg/sql/opt/xform/coster.go, line 1204 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66559#-McOaDEf2UCNn_dhCPMQ:-McPY2Ay5wFwh6j7ozxz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7259a1f3f35582a706ccf6d0bdcd066ad0ad1247/pkg/sql/opt/xform/coster.go#L1204)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] maybe include a comment explaining how to think about seqIOCostFactor vs randIOCostFactor here. The reality is that in most cases we're not simply appending rows, we're inserting into a sorted structure. But there is inherent batching in the LSM tree, so seqIOCostFactor still seems like the better choice.
</blockquote></details>

Done.
___
*[pkg/sql/opt/xform/coster.go, line 1205 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66559#-McOU3um-ntOh6J6Gng_:-McpmOd43UNUpiARVfpm:b-jsrqjm) ([raw file](https://github.com/cockroachdb/cockroach/blob/7259a1f3f35582a706ccf6d0bdcd066ad0ad1247/pkg/sql/opt/xform/coster.go#L1205)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It might be easier to reason about if we just made it a linear ramp between a "low" and a "high" cutoff, which would be a pretty smooth transition if there is significant difference between the cutoffs. It would be much cheaper to compute too.
</blockquote></details>

Done. Great idea!
___
*[pkg/sql/opt/xform/testdata/coster/groupby, line 28 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66559#-McMWHvz22FPzNLD7lFN:-McOzoge8C4cTsQRSl0l:byk6v8a) ([raw file](https://github.com/cockroachdb/cockroach/blob/7259a1f3f35582a706ccf6d0bdcd066ad0ad1247/pkg/sql/opt/xform/testdata/coster/groupby#L28)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: explain what these test results are supposed to show. Maybe something like: "The group by's individual cost (the difference in the group by's cost and the group by's input cost) follows a logistic curve. The difference in the group by's individual cost between 100k and 150k rows is greater than the differences between 50k and 100k rows, and 150k and 200k rows." :shrug: 

I threw the numbers into a spreadsheet and plotted the cost to verify that they do indeed follow a logistic function curve. It would be nice to make it easier to verify these test results, but I don't see any way to do that.



</blockquote></details>

Haha I also put the numbers in a spreadsheet to plot them. I couldn't think of a way to visualize it with our current testing framework, but thinking about it more, I realized that we could take advantage of Justin's opbench framework for this sort of testing, which also allows us to compare estimated to actual cost. I've removed the coster tests in favor of more opbench tests, which are already in csv format and can be readily imported into a spreadsheet for visualization. 

And what's really exciting is it seems like this costing actually sort of maps to reality!!
I put the results from the hash set opbench test in a spreadsheet and got this result (it's showing increasing numbers of rows on the x axis, cost on the y axis, and lines for estimated and actual costs for an intersect operation with 1 and 2 columns): 


This shows that the estimated and actual costs have the same shape (I scaled the estimated costs so they would be on the same relative scale as the actual costs). It shows that we're not sufficiently accounting for the difference in number of columns, but we can handle that in a separate change.

I need to check the other operations (I think I might need to re-do group by...), but I wanted to update the PR with this one at least since I think it's pretty cool!

**Comment by rytaft** [comment]
___
*[pkg/sql/opt/xform/testdata/coster/groupby, line 28 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66559#-McMWHvz22FPzNLD7lFN:-McpvxmYBCgtX9buhn9p:bspoqe0) ([raw file](https://github.com/cockroachdb/cockroach/blob/7259a1f3f35582a706ccf6d0bdcd066ad0ad1247/pkg/sql/opt/xform/testdata/coster/groupby#L28)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Haha I also put the numbers in a spreadsheet to plot them. I couldn't think of a way to visualize it with our current testing framework, but thinking about it more, I realized that we could take advantage of Justin's opbench framework for this sort of testing, which also allows us to compare estimated to actual cost. I've removed the coster tests in favor of more opbench tests, which are already in csv format and can be readily imported into a spreadsheet for visualization. 

And what's really exciting is it seems like this costing actually sort of maps to reality!!
I put the results from the hash set opbench test in a spreadsheet and got this result (it's showing increasing numbers of rows on the x axis, cost on the y axis, and lines for estimated and actual costs for an intersect operation with 1 and 2 columns): 


This shows that the estimated and actual costs have the same shape (I scaled the estimated costs so they would be on the same relative scale as the actual costs). It shows that we're not sufficiently accounting for the difference in number of columns, but we can handle that in a separate change.

I need to check the other operations (I think I might need to re-do group by...), but I wanted to update the PR with this one at least since I think it's pretty cool!
</blockquote></details>

(Thinking about this more, I think some of this shape has to do with the way the test was constructed with specific input row counts for the left and right sides, but I think this is still a pretty good result)

**Review by mgartner — APPROVED** [review]
Cool! :lgtm: 

> Reviewed 15 of 15 files at r3, 15 of 15 files at r4, 42 of 42 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66559#-:-Mcq9ZaXACRvyvsKQu9u:b-dp13x5)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[pkg/sql/opt/opbench/cat.go, line 42 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66559#-Mcq6Xzl9VKpzqLOVk0r:-Mcq6Xzm7syidGpT4YSG:bbgyu8j) ([raw file](https://github.com/cockroachdb/cockroach/blob/daa6474f64fdf3aae0f5e265520b693a9d1554b0/pkg/sql/opt/opbench/cat.go#L42)):*
> ```Go
>         "name": "__auto__",
>         "null_count": 0,
>         "row_count": 5
> ```

How did you update these stats? Are there documented steps?
___
*[pkg/sql/opt/optgen/exprgen/testdata/groupby, line 7 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66559#-Mcq0zOF8HjnP6KkHZRR:-Mcq0zOF8HjnP6KkHZRS:b8e25zh) ([raw file](https://github.com/cockroachdb/cockroach/blob/daa6474f64fdf3aae0f5e265520b693a9d1554b0/pkg/sql/opt/optgen/exprgen/testdata/groupby#L7)):*
> ```
> expr
> (Root
> 	(GroupBy
> ```

nit: I think we usually use spaces instead of tabs in testdata files.
___
*[pkg/sql/opt/xform/coster.go, line 1109 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66559#-Mcq4aAmBRyC1fWFfRP1:-Mcq4aAmBRyC1fWFfRP2:b6ru18s) ([raw file](https://github.com/cockroachdb/cockroach/blob/daa6474f64fdf3aae0f5e265520b693a9d1554b0/pkg/sql/opt/xform/coster.go#L1109)):*
`hashCost` is zero when the all group by columns are ordered in the input (AKA a streaming group by), correct? It might be helpful to explain that here, or in the comment directly above where it states:

> Add a cost that reflects the use of a hash table - unless we are doing a streaming aggregation where all the grouping columns are ordered

This makes sense, but hard to follow in the code because it's not obvious when we are doing a streaming aggregation. It's a leap to understand that it is streaming when `n == groupingColCount`.

**Review by RaduBerinde — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66559#-:-Mcr4O6iD9Q4mG8Qe7fJ:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[pkg/sql/opt/xform/coster.go, line 1108 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66559#-Mcr3_-ZART7D2s_OZc6:-Mcr3_-ZART7D2s_OZc7:bfc4746) ([raw file](https://github.com/cockroachdb/cockroach/blob/daa6474f64fdf3aae0f5e265520b693a9d1554b0/pkg/sql/opt/xform/coster.go#L1108)):*
> ```Go
> 		// n = 0:                factor = 1
> 		// n = groupingColCount: factor = 0
> 		hashCost *= 1 - memo.Cost(n)/memo.Cost(groupingColCount)
> ```

I don't think this "interpolation" makes sense any more - it did with the "hybrid" row operator that could hash part of the columns and stream the rest. Now it's either all hash or all streaming. I'd just have a case for `groupingColCount < n` where we add the hashing and buffering cost.
___
*[pkg/sql/opt/xform/testdata/coster/groupby, line 28 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66559#-McMWHvz22FPzNLD7lFN:-Mcr4Ogh7IUgBXhJ5v2_:b-uvozhf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7259a1f3f35582a706ccf6d0bdcd066ad0ad1247/pkg/sql/opt/xform/testdata/coster/groupby#L28)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

(Thinking about this more, I think some of this shape has to do with the way the test was constructed with specific input row counts for the left and right sides, but I think this is still a pretty good result)
</blockquote></details>

Very cool! It's a bit surprising it does look like that, I'd have expected the slope (i.e. the cost per row) to increase after the inflection point rather than decrease.

**Review by rytaft — COMMENTED** [review]
TFTRs! Updated.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66559#-:-McubJv82v_9rQjPayfV:b-u8hs9c)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner and @RaduBerinde)

___
*[pkg/sql/opt/opbench/cat.go, line 42 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66559#-Mcq6Xzl9VKpzqLOVk0r:-MctGD_8EXk666lmkMCQ:b-xo04xf) ([raw file](https://github.com/cockroachdb/cockroach/blob/daa6474f64fdf3aae0f5e265520b693a9d1554b0/pkg/sql/opt/opbench/cat.go#L42)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

How did you update these stats? Are there documented steps?
</blockquote></details>

I just used `EXPLAIN (opt, env)` with a query that cross joined all tables so I wouldn't get the histograms. Added a comment.
___
*[pkg/sql/opt/optgen/exprgen/testdata/groupby, line 7 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66559#-Mcq0zOF8HjnP6KkHZRR:-MctHWI38OMKbAkcisWw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/daa6474f64fdf3aae0f5e265520b693a9d1554b0/pkg/sql/opt/optgen/exprgen/testdata/groupby#L7)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: I think we usually use spaces instead of tabs in testdata files.
</blockquote></details>

Done.
___
*[pkg/sql/opt/xform/coster.go, line 1108 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66559#-Mcr3_-ZART7D2s_OZc6:-McubHq41Xwh6UXrtyVU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/daa6474f64fdf3aae0f5e265520b693a9d1554b0/pkg/sql/opt/xform/coster.go#L1108)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I don't think this "interpolation" makes sense any more - it did with the "hybrid" row operator that could hash part of the columns and stream the rest. Now it's either all hash or all streaming. I'd just have a case for `groupingColCount < n` where we add the hashing and buffering cost.
</blockquote></details>

Done.
___
*[pkg/sql/opt/xform/coster.go, line 1109 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66559#-Mcq4aAmBRyC1fWFfRP1:-McubFjd0gN9J_DkihyA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/daa6474f64fdf3aae0f5e265520b693a9d1554b0/pkg/sql/opt/xform/coster.go#L1109)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

`hashCost` is zero when the all group by columns are ordered in the input (AKA a streaming group by), correct? It might be helpful to explain that here, or in the comment directly above where it states:

> Add a cost that reflects the use of a hash table - unless we are doing a streaming aggregation where all the grouping columns are ordered

This makes sense, but hard to follow in the code because it's not obvious when we are doing a streaming aggregation. It's a leap to understand that it is streaming when `n == groupingColCount`.

</blockquote></details>

Done.
___
*[pkg/sql/opt/xform/testdata/coster/groupby, line 28 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66559#-McMWHvz22FPzNLD7lFN:-McseUo79Yr7pIMVWvwF:b-ebjewx) ([raw file](https://github.com/cockroachdb/cockroach/blob/7259a1f3f35582a706ccf6d0bdcd066ad0ad1247/pkg/sql/opt/xform/testdata/coster/groupby#L28)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Very cool! It's a bit surprising it does look like that, I'd have expected the slope (i.e. the cost per row) to increase after the inflection point rather than decrease.
</blockquote></details>

There are actually 3 inflection points in the lines for estimated cost, but it's pretty hard to tell. The first one is close to 0 -- we start the ramp at 64 rows (`noSpillRowCount`). 

There is a second, much more obvious one at around 3M rows, which is related to row counts, not the buffer cost. This is, as I alluded to, due to the way I set up this particular test. The number of rows on the left side of the `INTERSECT` increases according to the x-axis. These are rows from the `lineitem` table. But the number of rows on the right side of the `INTERSECT` only increases until the `orders` table is exhausted, which happens at 3M rows (for SF=2). After that point the right side is a full table scan of `orders`.

The last (very subtle) inflection point is between 6M and 7M, since we end the ramp at 6.4M (`spillRowCount`). It doesn't look like the lines for actual cost have an inflection point here (which would be very surprising if they did...), but it seems like they have similar asymptotic behavior.

So it's not quite as exciting as I first thought, but it's a big improvement over the old estimated cost without the buffer cost (the bottom lines in this chart here):

**Comment by RaduBerinde** [comment]
Thanks for those charts, very cool!
